### PR TITLE
Migrate from Buidler to Hardhat except docs

### LINKED ISSUE #206: Migrate Buidler plugin to Hardhat
Due to the [new way compilation works](https://hardhat.org/guides/compile-contracts.html), this is not as trivial as making the name change.

### PR DESCRIPTION
**Pull request by martriay**
Fixes #206

### CODE REVIEW

File: packages/plugin-hardhat/README.md [hunk 1]
```diff
@@ -1,39 +1,24 @@
...
-npm install --save-dev @openzeppelin/buidler-upgrades
-npm install --save-dev @nomiclabs/buidler-ethers ethers # peer dependencies
-```
-
-And register the plugins in your [`buidler.config.js`](https://buidler.dev/config/):
-
-```js
-usePlugin('@nomiclabs/buidler-ethers');
-usePlugin('@openzeppelin/buidler-upgrades');
-```
```
Comment (frangio) [inline packages/plugin-hardhat/README.md:20]: What are your thoughts on this section and why did you remove it?

(The TypeScript part that follows this is indeed no longer necessary.)

File: packages/plugin-hardhat/README.md [hunk 2]
```diff
@@ -1,39 +1,24 @@
...
 
 ## Usage in scripts
 
-You can use this plugin in a [Buidler script](https://buidler.dev/guides/scripts.html) to deploy an upgradeable instance of one of your contracts via the `deployProxy` function:
+You can use this plugin in a [Hardhat script](https://hardhat.org/guides/scripts.html) to deploy an upgradeable instance of one of your contracts via the `deployProxy` function:
 
 ```js
 // scripts/create-box.js
-const { ethers, upgrades } = require("@nomiclabs/buidler");
+const { ethers, upgrades } = require("@nomiclabs/hardhat");
```
Comment (frangio) [inline packages/plugin-hardhat/README.md:21]:
```suggestion
const { ethers, upgrades } = require("hardhat");
```

File: packages/plugin-hardhat/README.md [hunk 3]
```diff
@@ -51,7 +36,7 @@ Then, in another script, you can use the `upgradeProxy` function to upgrade the
 
 ```js
 // scripts/upgrade-box.js
-const { ethers, upgrades } = require("@nomiclabs/buidler");
+const { ethers, upgrades } = require("@nomiclabs/hardhat");
```
Comment (frangio) [inline packages/plugin-hardhat/README.md:39]:
```suggestion
const { ethers, upgrades } = require("hardhat");
```

File: packages/plugin-hardhat/test/linked-libraries.js [hunk 4]
```diff
@@ -105,7 +104,7 @@ async function deployLibrary(libraryName) {
 }
 
 async function getLinkedContractFactory(contractName, libraries) {
-  const cArtifact = await readArtifact(config.paths.artifacts, contractName);
+  const cArtifact = await artifacts.readArtifact(config.paths.artifacts, contractName);
```
Comment (frangio) [inline packages/plugin-hardhat/test/linked-libraries.js:107]: I don't think the path to artifacts is necessary now. (Based on [Migrating a plugin from Buidler to hardhat](https://hardhat.org/advanced/migrating-buidler-plugin.html).)

```suggestion
  const cArtifact = await artifacts.readArtifact(contractName);
```

File: .eslintrc.js [hunk 5]
```diff
@@ -19,7 +19,7 @@ module.exports = {
       extends: ['eslint:recommended', 'plugin:@typescript-eslint/recommended'],
     },
     {
-      files: ['buidler.config.js'],
+      files: ['hardhat.config.js'],
       globals: {
         usePlugin: 'readonly',
       },
```
Comment (frangio) [inline .eslintrc.js:25]: This can be removed because `usePlugin` is not used any more.

File: packages/core/src/validate.ts [hunk 6]
```diff
@@ -10,9 +10,10 @@ import { UpgradesError, ErrorDescriptions } from './error';
 import { SrcDecoder } from './src-decoder';
 import { isNullish } from './utils/is-nullish';
 
-export type Validation = Record<string, ValidationResult>;
+export type ValidationLog = Array<RunValidation>;
```
Comment (frangio) [inline packages/core/src/validate.ts:13]: These are equivalent but we use this syntax.

```suggestion
export type ValidationLog = RunValidation[];
```

File: packages/core/src/validate.ts [hunk 7]
```diff
@@ -120,57 +121,75 @@ export function validate(solcOutput: SolcOutput, decodeSrc: SrcDecoder): Validat
...
     throw new Error(`Contract ${contractName} is abstract`);
   }
   return version;
 }
 
-function getContractName(validation: Validation, version: Version): string {
-  const contractName = Object.keys(validation).find(
-    name => validation[name].version?.withMetadata === version.withMetadata,
-  );
+function getContractNameAndCompilerRun(validations: ValidationLog, version: Version): [string, number] {
```
Comment (frangio) [inline packages/core/src/validate.ts:132]: I think this would be better.

```suggestion
function getContractNameAndRunValidation(validations: ValidationLog, version: Version): [string, RunValidation] {
```

File: packages/core/src/validate.ts [hunk 8]
```diff
@@ -120,57 +121,75 @@ export function validate(solcOutput: SolcOutput, decodeSrc: SrcDecoder): Validat
...
-    Object.assign(layout.types, validation[name].layout.types);
+    layout.storage.unshift(...validations[compilerRun][name].layout.storage);
+    Object.assign(layout.types, validations[compilerRun][name].layout.types);
   }
   return layout;
 }
 
-export function getUnlinkedBytecode(validation: Validation, bytecode: string): string {
-  const linkableContracts = Object.keys(validation).filter(name => validation[name].linkReferences.length > 0);
+export function getUnlinkedBytecode(validations: ValidationLog, bytecode: string): string {
```
Comment (frangio) [inline packages/core/src/validate.ts:164]: This function wasn't updated correctly. It's meant to look through all of the available validations. It should look through the entire validation log.

I think the correct solution is to iterate over each `validation` in the log running the same code we had before.

```suggestion
for (const validation of ValidationLog) {
  // the body of the function as it was before this PR
}
```

File: packages/plugin-hardhat/scripts/test.sh [hunk 9]
```diff
@@ -3,5 +3,5 @@
 set -euo pipefail
 
 rimraf .openzeppelin
-buidler compile
+npx hardhat compile
```
Comment (frangio) [inline packages/plugin-hardhat/scripts/test.sh:6]: This shouldn't be necessary, the script should be run via `npm test`.

```suggestion
hardhat compile
```

File: packages/plugin-hardhat/src/validations.ts [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+  await fs.mkdir(hre.config.paths.cache, { recursive: true });
+  await fs.writeFile(getValidationsCachePath(hre), JSON.stringify(validations, null, 2));
+}
+
+export async function readValidations(hre: HardhatRuntimeEnvironment): Promise<ValidationLog> {
+  try {
+    return JSON.parse(await fs.readFile(getValidationsCachePath(hre), 'utf8'));
+  } catch (e) {
+    if (e.code === 'ENOENT') {
+      throw new Error('ValidationLog cache not found. Recompile with `hardhat compile --force`');
```
Comment (frangio) [inline packages/plugin-hardhat/src/validations.ts:30]:
```suggestion
      throw new Error('Validations log not found. Recompile with `hardhat compile --force`');
```

File: packages/plugin-hardhat/src/index.ts [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+import { TASK_COMPILE_SOLIDITY_COMPILE } from 'hardhat/builtin-tasks/task-names';
+import { lazyObject } from 'hardhat/plugins';
+import { validate, solcInputOutputDecoder, SolcInput } from '@openzeppelin/upgrades-core';
+import { writeValidations } from './validations';
+
+interface RunCompilerArgs {
+  input: SolcInput;
+}
+
+subtask(TASK_COMPILE_SOLIDITY_COMPILE, async (args: RunCompilerArgs, hre, runSuper) => {
```
Comment (frangio) [inline packages/plugin-hardhat/src/index.ts:15]: We have to check whether Harhdat can run multiple compilations concurrently. In this case there is a race condition with `writeValidations`. Not sure yet how to solve it if it is indeed a problem.

**Review by frangio — COMMENTED** [review]
Thanks!

File: packages/plugin-hardhat/src/index.ts (hunk 11 above)
Comment (frangio) [reply packages/plugin-hardhat/src/index.ts:15]: Just confirmed with Hardhat team that compilations don't run concurrently now, but may in the future.

File: packages/plugin-hardhat/scripts/test.sh (hunk 9 above)
Comment (martriay) [reply packages/plugin-hardhat/scripts/test.sh:6]: oops, this was a debugging attempt it shouldn't be here!

File: packages/core/src/validate.ts (hunk 7 above)
Comment (martriay) [reply packages/core/src/validate.ts:132]: Mhh I agree that we can do better but we already have a `RunValidation` type and this function isn't returning one. What about `getContractNameAndRunValidationId`? or `getContractNameAndValidationId` for short.

File: packages/plugin-hardhat/README.md (hunk 1 above)
Comment (martriay) [reply packages/plugin-hardhat/README.md:20]: I removed it by mistake, adding it back now.

File: assets/hardhat.svg [hunk 12]
```diff
@@ -0,0 +1 @@
...,13.92s1.93,14.08,11.26,14.08c9.57,0,11.58-5.63,11.58-14.08C158,264.39,156.3,258.43,146.49,258.43Z"/><path class="a8715c47-c5e1-4023-9511-db57a031c855" d="M206.34,292.94V269.86c0-7.73-1.29-11.43-10.06-11.43s-10.54,3.62-10.54,11.43l-.16,23.08h-8.44V236.07h8.44v20.84c1.61-3,5.23-5.72,13.12-5.72,13.43,0,16.17,7.65,16.17,19.47v22.28Z"/><path class="a8715c47-c5e1-4023-9511-db57a031c855" d="M257.18,292.94l-1.77-6.19c-1.93,3.22-7.64,6.76-16.17,6.76-8.85,0-15.93-3.94-15.93-12.63,0-7.56,5.39-11,17.14-12.15l13-1.29c-.24-5.95-2.42-9.41-10.7-9.41-7.64,0-9.74,3.06-9.74,8.21l-8.36.08c0-8.53,3.54-15.13,18-15.13,16.81,0,19.22,7.24,19.22,18.35v23.4Zm-3.78-19-11.67,1.29c-7.16.8-9.57,2.33-9.57,5.55,0,3.78,3.22,5.95,9.25,5.95,7.56,0,12-4.34,12-11.5Z"/><path class="a8715c47-c5e1-4023-9511-db57a031c855" d="M281.55,258.67v22.45c0,3.86,1.61,5.63,10.54,5.23v6.92c-13.59.8-19.3-2.42-19.3-12.15V258.67h-5.88v-5.14l5.88-1.77v-10.3h8.52v10.3h10.78v6.91Z"/></g><use width="292" height="204.4" transform="translate(0.28)" xlink:href="#f2725a89-cd96-4a0d-bc31-31cd625566bf"/></symbol></defs><use width="292.28" height="293.51" transform="translate(50.94 51.49)" xlink:href="#a6eb6c2d-e035-4ab2-bb19-8be78e9e949e"/></svg>
```
Comment (frangio) [inline assets/hardhat.svg:1]: This icon looks better: https://hardhat.org/apple-touch-icon.png

File: packages/plugin-hardhat/README.md [hunk 13]
```diff
@@ -12,13 +12,25 @@ npm install --save-dev @openzeppelin/hardhat-upgrades
...
+
+```js
+// Javascript
+require('@nomiclabs/hardhat-ethers');
+require('@openzeppelin/hardhat-upgrades');
+
+// Typescript
+import '@nomiclabs/hardhat-ethers';
+import '@openzeppelin/hardhat-upgrades';
+```
```
Comment (frangio) [inline packages/plugin-hardhat/README.md]: Ethers plugin is implicit.

```suggestion
And register the plugin in your [`hardhat.config.js`](https://hardhat.org/config/):

```js
// JavaScript
require('@openzeppelin/hardhat-upgrades');

// TypeScript
import '@openzeppelin/hardhat-upgrades';
```
```

**Review by frangio — APPROVED** [review]
Thank you!
